apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.CommonTesting/SpannerTestDatabaseBase.cs

```diff
@@ -0,0 +1,224 @@
+
+    /// <summary>
+    /// The Spanner Host name to connect to. It is read from the environment variable "TEST_SPANNER_HOST".
+    /// </summary>
+    public virtual string SpannerHost { get; } = GetEnvironmentVariableOrDefault("TEST_SPANNER_HOST", null);
+
+    /// <summary>
+    /// The Spanner port number to connect to. It is read from the environment variable "TEST_SPANNER_PORT".
+    /// </summary>
+    public virtual string SpannerPort { get; } = GetEnvironmentVariableOrDefault("TEST_SPANNER_PORT", null);
+
+    /// <summary>
+    /// The Spanner Instance name to connect to. It is read from the environment variable "TEST_SPANNER_INSTANCE".
+    /// </summary>
+    public virtual string SpannerInstance { get; } = GetEnvironmentVariableOrDefault("TEST_SPANNER_INSTANCE", "spannerintegration");
+
+    /// <summary>
+    /// The Spanner Database name to connect to. It is read from the environment variable "TEST_SPANNER_DATABASE".
+    /// The derived class must provide its implementation to return the database name.
+    /// </summary>
+    public abstract string SpannerDatabase { get; }
+
+    /// <summary>
+    /// Returns true if the database was created just for this test, or false if the database was an existing one
+    /// specified through an environment variable.
+    /// </summary>
+    public abstract bool Fresh { get; }
+
+    /// <summary>
+    /// Connection string including database.
+    /// </summary>
+    public string ConnectionString { get; }
+
+    /// <summary>
+    /// Connection string without the database.
+    /// </summary>
+    public string NoDbConnectionString { get; }
+
+    /// <summary>
+    /// The GCP Project ID to be used for the test database.
+    /// </summary>
+    public string ProjectId { get; }
+
+    /// <summary>
+    /// The fully-qualified database name.
+    /// </summary>
+    public DatabaseName DatabaseName { get; }
+
+    internal SpannerClientCreationOptions SpannerClientCreationOptions { get; }
+
+    protected SpannerTestDatabaseBase(string projectId, DatabaseDialect dialect)
+    {
+        TestLogger.Install();
+        ProjectId = projectId;
+        var builder = new SpannerConnectionStringBuilder
+        {
+            Host = SpannerHost,
+            DataSource = $"projects/{ProjectId}/instances/{SpannerInstance}",
+            EmulatorDetection = dialect == DatabaseDialect.Postgresql
+                ? EmulatorDetection.None
+                : EmulatorDetection.EmulatorOrProduction
+        };
+        if (SpannerPort != null)
+        {
+            builder.Port = int.Parse(SpannerPort);
+        }
+        NoDbConnectionString = builder.ConnectionString;
+        SpannerClientCreationOptions = new SpannerClientCreationOptions(builder);
+        var databaseBuilder = builder.WithDatabase(SpannerDatabase);
+        ConnectionString = databaseBuilder.ConnectionString;
+        DatabaseName = databaseBuilder.DatabaseName;
+
+        if (dialect != DatabaseDialect.Postgresql)
```

Comment:
Here as well you can ask about the EmulatorDetection value received in the constructor and not the dialect.

Reply:
Actually, you probably don't even need to ask anything here, as `MaybeCreateInstanceOnEmulator` is already asking about `SpannerClientCreationOptions.UsesEmulator` which I think would reflect whatever you set on the connection builder?

Reply:
Right Amanda. I did two things incorrectly. One was conditional creation in the base class and setting emulator detection based on dialect. I didn't understood EmulatorDetection correctly. I will make the change in the next commit.

Reply:
Sounds good, thanks!